### PR TITLE
OperationContext

### PR DESCRIPTION
**Pull request by rozza**
The OperationContext class is now a holder for:

 - Operation Id
 - `RequestContext`
 - `SessionContext`
 - `TimeoutContext`

The `BindingContext` interface has been simplified to only have an `OperationContext` getter. Elsewhere `OperationContext` is supplied down through the code so its accessible.

Note: 
* The Operation currently supplies the `TimeoutSettings` this maybe further refactored to be calculated at the `OperationExecutor` level (JAVA-5176).

### CODE REVIEW

**Comment by rozza** [comment]
Build: https://spruce.mongodb.com/version/6512f8efd1fe07c300f6df36/

File: driver-core/src/main/com/mongodb/internal/operation/WriteOperation.java [hunk 1]
```diff
@@ -24,6 +25,12 @@
  * <p>This class is not part of the public API and may be removed or changed at any time</p>
  */
 public interface WriteOperation<T> {
+
+    /**
+     * @return the timeout settings for this operation
+     */
+    TimeoutSettings getTimeoutSettings();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/WriteOperation.java:32]: I think we talked about removing this from Read/WriteOperation and instead passing it directly to the operation executor. Did we reject that idea or is it still on the table?

File: driver-core/src/test/functional/com/mongodb/internal/binding/AsyncOperationContextBinding.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import static org.bson.assertions.Assertions.notNull;
+
+public final class AsyncOperationContextBinding implements AsyncReadWriteBinding {
+
+    private final AsyncReadWriteBinding wrapped;
+    private final OperationContext operationContext;
+
+    public AsyncOperationContextBinding(final AsyncReadWriteBinding wrapped, final OperationContext operationContext) {
+        this.wrapped = notNull("wrapped", wrapped);
+        this.operationContext = operationContext;
```
Comment (katcharov) [inline driver-core/src/test/functional/com/mongodb/internal/binding/AsyncOperationContextBinding.java:34]: Why not both notNull?

File: driver-core/src/test/functional/com/mongodb/internal/binding/OperationContextBinding.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Override
+    public int getCount() {
+        return wrapped.getCount();
+    }
+
+    @Override
+    public ReadWriteBinding retain() {
+        wrapped.retain();
```
Comment (katcharov) [inline driver-core/src/test/functional/com/mongodb/internal/binding/OperationContextBinding.java:47]: Minor, but is it worth having a test for this class (the retain/release behaviour)?

File: driver-core/src/main/com/mongodb/internal/operation/WriteOperation.java (hunk 1 above)
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/operation/WriteOperation.java:32]: I think the plan is to do it in https://jira.mongodb.org/browse/JAVA-5176

**Review by katcharov — CHANGES_REQUESTED** [review]
It's looking good

File: driver-core/src/main/com/mongodb/internal/operation/SyncOperationHelper.java [hunk 4]
```diff
@@ -275,10 +290,16 @@ static <D, T> T createReadCommandAndExecute(
             final Decoder<D> decoder,
             final CommandReadTransformer<D, T> transformer,
             final Connection connection) {
-        BsonDocument command = commandCreator.create(clientSideOperationTimeout, source.getServerDescription(), connection.getDescription());
+        BsonDocument command = commandCreator.create(binding.getOperationContext(), source.getServerDescription(),
+                connection.getDescription());
         retryState.attach(AttachmentKeys.commandDescriptionSupplier(), command::getFirstKey, false);
-        return transformer.apply(assertNotNull(connection.command(database, command, new NoOpFieldNameValidator(),
-                source.getReadPreference(), decoder, binding)), source, connection);
+        // TODO REVERT
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/SyncOperationHelper.java:296]: Fixed

File: driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java [hunk 5]
```diff
@@ -400,14 +403,16 @@ private boolean handleMongoWriteConcernWithResponseExceptionAsync(final RetrySta
     @Nullable
     private BsonDocument executeCommand(final Connection connection, final BulkWriteBatch batch, final WriteBinding binding) {
         return connection.command(namespace.getDatabaseName(), batch.getCommand(), NO_OP_FIELD_NAME_VALIDATOR, null, batch.getDecoder(),
-                binding, shouldAcknowledge(batch, binding.getSessionContext()), batch.getPayload(), batch.getFieldNameValidator());
+                binding.getOperationContext(), shouldAcknowledge(batch, binding.getOperationContext().getSessionContext()),
+                batch.getPayload(), batch.getFieldNameValidator());
     }
 
+    // TODO - just pass in opContext
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java:410]: Fixed

File: driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java [hunk 6]
```diff
@@ -254,16 +271,16 @@ static <T, R> void executeRetryableWriteAsync(
...
-                                transformingWriteCallback(transformer, connection, addingRetryableLabelCallback));
+                        connection.commandAsync(database, command, fieldNameValidator, readPreference, commandResultDecoder,
+                                operationContext, transformingWriteCallback(transformer, connection, addingRetryableLabelCallback));
                     });
         }).whenComplete(binding::release);
 
         asyncWrite.get(exceptionTransformingCallback(errorHandlingCallback(callback, OperationHelper.LOGGER)));
     }
 
+    // TODO just pass OPContext
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java:282]: Fixed

File: driver-core/src/main/com/mongodb/internal/connection/CommandProtocolImpl.java [hunk 7]
```diff
@@ -95,13 +86,13 @@ public void executeAsync(final InternalConnection connection, final SingleResult
 
     @Override
     public CommandProtocolImpl<T> sessionContext(final SessionContext sessionContext) {
-        this.sessionContext = sessionContext;
+        this.operationContext = operationContext.withSessionContext(sessionContext);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/CommandProtocolImpl.java]: SGTM! done

**Comment by rozza** [comment]
New build: https://spruce.mongodb.com/version/651430781e2d17297de7c91e/

**Review by katcharov — APPROVED** [review]
LGTM

**Comment by rozza** [comment]
Closed - manually merged / folded and squashed.

**Pull request closed without merging**